Caffeinate command

ysmaliak opened the pull request:
## Description

Script command. Starts a caffeinated session or stops the current one. Support for 2 arguments: the first is the command (enable/e, disable/d), the second is the time in s, m, h, or d (e. g. 1s, 1m, 1h, 1d).

## Type of change

- [x] New script command

## Screenshot

<img width="1013" alt="Screenshot 2021-06-19 at 14 38 40" src="https://user-images.githubusercontent.com/42275599/122641158-1a8a0e00-d10c-11eb-90d7-4cc7847775bf.png">

## Checklist

- [x] I have read [Contribution Guidelines](https://github.com/raycast/script-commands/blob/master/CONTRIBUTING.md)

es183923 commented:
It would make more sense to split this into two commands instead of one, so that one command enables caffeinate, the other command disables caffeinate. (This is just my personal opinion, and I do not work for Raycast.)

ysmaliak commented:
> It would make more sense to split this into two commands instead of one, so that one command enables caffeinate, the other command disables caffeinate. (This is just my personal opinion, and I do not work for Raycast.)

It looks like a good idea, but then there will be a need to divide the script into 2 files, which do not make sense without each other (or have but little, hardly anyone wants to just turn off the caffeinate). Because of that, I decided to keep both commands under one script command

dehesa commented:
Hey @ysmaliak, thank you for the PR. It is good stuff!

I also quite liked you were able to circumvent the asynchrony problem ([check this out](https://github.com/raycast/script-commands/pull/141)). Great work!

As @es183923 mentions, we do prefer to have several commands. It is not a hard rule, but we encourage it. Please make it two commands, or even better, pick also the idea of listing cafffeinate processes from #141 .

ysmaliak commented:
Hey @dehesa, I split the command into separate commands and also added 2 commands to show caffeinate status. The one written in bash is pretty simple and quick, the other one written in Swift is more complex and also showing time left alongside with status, though not sure if it's a good idea to keep it `inline` due to its performance

ysmaliak commented:
<img width="824" alt="Screenshot 2021-06-23 at 20 43 57" src="https://user-images.githubusercontent.com/42275599/123144221-20f9ec00-d464-11eb-8b5d-d484a67b0da3.png">
<img width="819" alt="Screenshot 2021-06-23 at 20 44 23" src="https://user-images.githubusercontent.com/42275599/123144265-2e16db00-d464-11eb-8c6b-6c99bfbc28f0.png">
<img width="818" alt="Screenshot 2021-06-23 at 20 44 40" src="https://user-images.githubusercontent.com/42275599/123144289-366f1600-d464-11eb-8240-189377730522.png">
<img width="816" alt="Screenshot 2021-06-23 at 20 45 33" src="https://user-images.githubusercontent.com/42275599/123144326-3ec75100-d464-11eb-8098-277cfe0edae1.png">

dehesa commented:
Thanks for all the changes. It already looks great.

The performance of the Swift one is not great the first time, although it later on get cached. Are you sure you cannot make the shell one work more like [this one](https://github.com/maximkrouk/script-commands/blob/d4b4f7d29ebecbebe60504827dd12d891b212411/commands/system/longest-caffeinate.sh)? It does look pretty scary 😆 

I left one comment in the Swift file.

dehesa commented on commands/system/caffeinate-status.swift:
```diff
@@ -0,0 +1,97 @@
+if let maxTime = maxTime, let currentTimeFromFormatter = currentTimeFromFormatter {
+    let (d, h, m) = secondsToDhm(Int(maxTime - currentTimeFromFormatter))
+    let hmString: String = (d > 0 ? "\(d)d " : "") + (h > 0 ? "\(h)h " : "") + (m > 0 ? "\(m)m" : "")
+    print("Caffeinate is running. Time left: \(hmString)")
```
A user will see:
```
☕ Caffeinate  Caffeinate is running. Time left: 21 m
```

It feels a bit repetitive. What about?
```
☕ Caffeinate  Started at 16:45. Expires in 30m.
```

ysmaliak replied:
I've also added a date in case a user sets caffeinate for a few days, so he won't be confused by time without a date (the date will be shown only if the day when caffeinate was enabled is not today)

`Started at 14:19. Expires in 47m`
or
`Started on Jun 23, 14:19. Expires in 47m`
